### PR TITLE
Bubble Sort Exercise

### PR DESCRIPTION
**Pull request by kcisneros**

### CODE REVIEW

**Comment by kcisneros** [comment]
Mostly just pseudocode & doing the basic comparison first. I need to push the sorted elements into a new array and once that array doesn't have any sorts, return the new (final) array. Currently running into this problem:

```
10297-Katherine-Cisneros:bubble_sort kcisneros-study$ ruby bubble_sort.rb 
"number is: 4"
"index is: 0"
"next number is: 3"
4 is greater than 3, shift it!
"number is: 3"
"index is: 1"
"next number is: 78"
3 is not greater than 78, don't shift it!
"number is: 78"
"index is: 2"
"next number is: 2"
78 is greater than 2, shift it!
"number is: 2"
"index is: 3"
"next number is: 0"
2 is greater than 0, shift it!
"number is: 0"
"index is: 4"
"next number is: 2"
0 is not greater than 2, don't shift it!
"number is: 2"
"index is: 5"
"next number is: "
bubble_sort.rb:24:in `>': comparison of Integer with nil failed (ArgumentError)
```

I need to fix my loop to iterate again, seems it's only going through once..

**Comment by kcisneros** [comment]
Current print to terminal:

```
10297-Katherine-Cisneros:bubble_sort kcisneros-study$ ruby bubble_sort.rb 
"number is: 4"
"index is: 0"
"next number is: 3"
sorted_array is: [3, 4]
4 is greater than 3, shift it!
"number is: 3"
"index is: 1"
"next number is: 78"
3 is not greater than 78, don't shift it!
"number is: 78"
"index is: 2"
"next number is: 2"
sorted_array is: [3, 4, 2, 78]
78 is greater than 2, shift it!
"number is: 2"
"index is: 3"
"next number is: 0"
sorted_array is: [3, 4, 2, 78, 0, 2]
2 is greater than 0, shift it!
"number is: 0"
"index is: 4"
"next number is: 2"
0 is not greater than 2, don't shift it!
"number is: 2"
"index is: 5"
"next number is: "
bubble_sort.rb:24:in `>': comparison of Integer with nil failed (ArgumentError)
        from bubble_sort.rb:24:in `block in bubble_sort'
        from bubble_sort.rb:20:in `each'
        from bubble_sort.rb:20:in `each_with_index'
        from bubble_sort.rb:20:in `bubble_sort'
        from bubble_sort.rb:35:in `<main>'
```

My loop isn't quite right yet.. but some shifting is happening, which is good.. 👍

**Comment by kcisneros** [comment]
this was hard. ain't no way to say it. the three exercises in TOP (bubble, stocks, and dictionary) were incredibly difficult for me 😞 but, at least this one is done :) and onto more difficult things

File: ruby_basics/bubble_sort/bubble_sort.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  swapping = true 
+
+  while swapping 
+    swapping = false
+
+    array.each_with_index do |number, index|
+
+      # if the number or the next number are nil (end of array)
+      # immediately stop the code
+      if number.nil? || array[index+1].nil?
```
Comment (chmurph2) [inline ruby_basics/bubble_sort/bubble_sort.rb:19]: I don't think you need to check `number.nil?`, as I don't think that can ever happen. 

I removed it and this program still works.

File: ruby_basics/bubble_sort/bubble_sort.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        break   
+      end
+
+      # puts "swapping at the top is: #{swapping}"
+
+      # puts "number is: #{number}"
+      # puts "next num is #{array[index+1]}"
+      # puts "index is: #{index}"
+
```
Comment (chmurph2) [inline ruby_basics/bubble_sort/bubble_sort.rb:28]: I think lines 29 - 50 are unnecessarily indented.

File: ruby_basics/bubble_sort/bubble_sort.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          # swap the values in the array. aka 4, 3 becomes 3, 4 in the array
+          array[index], array[index+1] = array[index+1], array[index]
+          # puts "updated array in the if is: #{array}"
+
+        # if the above isn't true
+        else
+          # puts "#{number} is not greater than #{array[index+1]}, don't shift it!"
+          # leave the values in the array as they are. aka 3, 4 stays 3, 4
+          array[index+1], array[index] = array[index+1], array[index]
+          # puts "updated array in the else is: #{array}"
```
Comment (chmurph2) [inline ruby_basics/bubble_sort/bubble_sort.rb:43]: I don't think you need to do this at all. I commented it out and this program still works.

**Review by chmurph2 — APPROVED** [review]
Looks good overall. I left some suggestions on how you can make it even tighter by, I believe, removing some unnecessary code

**Pull request merged**